services/storehost/extStats.go

```diff
@@ -0,0 +1,455 @@
+		mClient metadata.TChanMetadataService
+		logger  bark.Logger
+
+		extents    extStatsContextMap
+		reportPool sync.Pool
+		reportC    chan *report
+
+		wg    sync.WaitGroup
+		stopC chan struct{} // internal channel to stop
+	}
+
+	extStatsContextMap struct {
+		sync.RWMutex
+		xmap map[string]*extStatsContext
+	}
+
+	extStatsContext struct {
+		ext        *extentContext
+		ref        int64
+		lastReport *report
+	}
+
+	report struct {
+		ctx       *extStatsContext
+		extentID  uuid.UUID
+		timestamp int64
+
+		firstSeqNum, lastSeqNum int64
+		sealed                  bool
+	}
+)
+
+// the following globals and their methods are
+var (
+	reportInterval         atomic.Value
+	extStatsReporterPaused int32 = 0
+)
+
+func init() {
+	reportInterval.Store(defaultReportInterval)
+}
+
+// ExtStatsReporterPause pauses the reporting (intended for tests)
+func ExtStatsReporterPause() {
+
+	swapped := atomic.CompareAndSwapInt32(&extStatsReporterPaused, 0, 1)
+	common.GetDefaultLogger().WithField(`was-running`, swapped).Info("extStatsReporter: paused")
+}
+
+// ExtStatsReporterResume resumes the reporting (intended for tests)
+func ExtStatsReporterResume() {
+
+	swapped := atomic.CompareAndSwapInt32(&extStatsReporterPaused, 1, 0)
+	common.GetDefaultLogger().WithField(`was-running`, !swapped).Info("extStatsReporter: resumed")
+}
+
+// ExtStatsReporterSetReportInterval updates the report interval (intended for tests)
+func ExtStatsReporterSetReportInterval(interval time.Duration) {
+
+	reportInterval.Store(interval)
+	common.GetDefaultLogger().WithField(`interval`, interval).Info("extStatsReporter: updated report interval")
+}
+
+func NewExtStatsReporter(hostID string, xMgr *ExtentManager, mClient metadata.TChanMetadataService, logger bark.Logger) *ExtStatsReporter {
+
+	return &ExtStatsReporter{
+		hostID:  hostID,
+		xMgr:    xMgr,
+		mClient: mClient,
+		logger:  logger,
+
+		extents: extStatsContextMap{
+			xmap: make(map[string]*extStatsContext),
+		},
+
+		reportPool: sync.Pool{New: func() interface{} { return &report{} }},
+		reportC:    make(chan *report, reportChanBufLen),
+		stopC:      make(chan struct{}),
+	}
+}
+
+func (t *ExtStatsReporter) Start() {
+
+	t.logger.Info("extStatsReporter: started")
+
+	// register callbacks to get notified every time an extent is opened/closed
+	t.xMgr.RegisterCallbacks(t)
+
+	t.wg.Add(1)
+	go t.reporterPump()
+
+	t.wg.Add(1)
+	go t.schedulerPump()
+}
+
+func (t *ExtStatsReporter) Stop() {
+
+	close(t.stopC)
+	t.wg.Wait()
+
+	t.logger.Info("extStatsReporter: stopped")
+}
+
+func (t *ExtStatsReporter) trySendReport(extentID uuid.UUID, ext *extentContext, ctx *extStatsContext) bool {
+
+	// if paused, do nothing
+	if atomic.LoadInt32(&extStatsReporterPaused) == 1 {
+		return false
+	}
+
+	r := t.reportPool.Get().(*report)
+
+	r.ctx = ctx
+	r.extentID = extentID
+	r.timestamp = time.Now().UnixNano()
+
+	// get a snapshot of various extent stats
+	r.firstSeqNum, r.lastSeqNum = ext.getBeginSeqNum(), ext.getLastSeqNum()
```

Comment:
yes, i have another diff in the works, that publishes start/end timestamps and addresses -- in order to fill out all the other fields in the extent-stats-report.